cpp/src/fil/fil.cu

```diff
@@ -333,37 +321,90 @@ void node2fil(std::vector<dense_node_t>* pnodes, int root, int cur,
       break;
     case tl::Operator::kLE:
       // x <= y is equivalent to x < y', where y' is the next representable float
-      threshold =
-        std::nextafterf(threshold, std::numeric_limits<float>::infinity());
+      *pthreshold =
+        std::nextafterf(*pthreshold, std::numeric_limits<float>::infinity());
       break;
     case tl::Operator::kGT:
       // x > y is equivalent to x >= y', where y' is the next representable float
       // left and right still need to be swapped
-      threshold =
-        std::nextafterf(threshold, std::numeric_limits<float>::infinity());
+      *pthreshold =
+        std::nextafterf(*pthreshold, std::numeric_limits<float>::infinity());
     case tl::Operator::kGE:
       // swap left and right
-      std::swap(left, right);
-      default_left = !default_left;
+      std::swap(*tl_left, *tl_right);
+      *default_left = !*default_left;
       break;
     default:
       ASSERT(false, "only <, >, <= and >= comparisons are supported");
   }
-  dense_node_init(&nodes[root + cur], 0, threshold, node.split_index(),
+}
+
+void node2fil_dense(std::vector<dense_node_t>* pnodes, int root, int cur,
+                    const tl::Tree& tree, const tl::Tree::Node& node) {
+  if (node.is_leaf()) {
+    dense_node_init(&(*pnodes)[root + cur], node.leaf_value(), 0, 0, false,
+                    true);
+    return;
+  }
+
+  // inner node
+  ASSERT(node.split_type() == tl::SplitFeatureType::kNumerical,
+         "only numerical split nodes are supported");
+  int tl_left = node.cleft(), tl_right = node.cright();
+  bool default_left = node.default_left();
+  float threshold = node.threshold();
+  adjust_threshold(&threshold, &tl_left, &tl_right, &default_left, node);
+  dense_node_init(&(*pnodes)[root + cur], 0, threshold, node.split_index(),
                   default_left, false);
-  node2fil(pnodes, root, 2 * cur + 1, tree, tl_node_at(tree, left));
-  node2fil(pnodes, root, 2 * cur + 2, tree, tl_node_at(tree, right));
+  int left = 2 * cur + 1;
+  node2fil_dense(pnodes, root, left, tree, tl_node_at(tree, tl_left));
+  node2fil_dense(pnodes, root, left + 1, tree, tl_node_at(tree, tl_right));
+}
+
+void node2fil_sparse(std::vector<sparse_node_t>* pnodes, int root, int cur,
+                     const tl::Tree& tree, const tl::Tree::Node& node) {
+  if (node.is_leaf()) {
+    sparse_node_init(&(*pnodes)[root + cur], node.leaf_value(), 0, 0, false,
+                     true, 0);
+    return;
+  }
+
+  // inner node
+  ASSERT(node.split_type() == tl::SplitFeatureType::kNumerical,
+         "only numerical split nodes are supported");
+  int tl_left = node.cleft(), tl_right = node.cright();
+  bool default_left = node.default_left();
+  float threshold = node.threshold();
+  adjust_threshold(&threshold, &tl_left, &tl_right, &default_left, node);
+
+  // reserve space for child nodes
+  int left = pnodes->size() - root;
```

Comment:
I confess to not understanding how left and tl_left differ. They both look like int offsets pointing to the left child node. Is one of them an offset wrt the tree root and one relative to the whole forest root?

Reply:
Maybe a comment would help?